File: app/src/main/java/com/pluu/support/nate/NateDetailApi.java
package com.pluu.support.nate;

import java.util.ArrayList;
import java.util.List;

import com.pluu.support.impl.AbstractDetailApi;
import com.pluu.webtoon.item.Detail;
import com.pluu.webtoon.item.DetailView;
import com.pluu.webtoon.item.Episode;
import com.pluu.webtoon.item.ShareItem;
import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;

/**
 * 네이트 웹툰 상세 API
 * Created by PLUUSYSTEM-NEW on 2015-10-31.
 */
public class NateDetailApi extends AbstractDetailApi {

	private final String DETAIL_URL = "http://m.comics.nate.com/view/show?btno=%s&bsno=%s&order=up";
	private String webToonId, episodeId;

	@Override
	public Detail parseDetail(Episode episode) {
		this.webToonId = episode.getWebtoonId();
		this.episodeId = episode.getEpisodeId();

		Detail ret = new Detail();
		ret.webtoonId = episode.getWebtoonId();

		String response;
		try {
			response = requestApi();
		} catch (Exception e) {
			e.printStackTrace();
			ret.list = new ArrayList<>();
			return ret;
		}

		Document doc = Jsoup.parse(response);
		ret.title = doc.select(".tvi_header").text();
		ret.episodeId = episodeId;

		Elements temp = doc.select(".btn_prev");
		if (temp != null && !temp.isEmpty()) {
			ret.prevLink = "/view/" + temp.first().attr("href");
		}

		temp = doc.select(".btn_next");
		if (temp != null && !temp.isEmpty()) {
			ret.nextLink = "/view/" + temp.first().attr("href");
		}

		List<DetailView> list = new ArrayList<>();
		Elements elements = doc.select(".fview");
		for (Element img : elements) {
			list.add(DetailView.createImage(img.attr("src")));
		}

		ret.list = list;
		return ret;
	}

	@Override
	public ShareItem getDetailShare(Episode episode, Detail detail) {
		ShareItem item = new ShareItem();
		item.title = episode.getTitle() + " / " + detail.title;
		item.url = String.format(DETAIL_URL, episode.getWebtoonId(), episode.getEpisodeId());
		return item;
	}

	@Override
	public String getMethod() {
		return GET;
	}

	@Override
	public String getUrl() {
		return String.format(DETAIL_URL, webToonId, episodeId);
	}

}

File: app/src/main/java/com/pluu/support/nate/NateDetailApi.kt
package com.pluu.support.nate

import com.pluu.support.impl.AbstractDetailApi
import com.pluu.support.impl.REQUEST_METHOD
import com.pluu.webtoon.di.NetworkUseCase
import com.pluu.webtoon.item.*
import org.jsoup.Jsoup

/**
 * 네이트 웹툰 상세 API
 * Created by pluu on 2017-04-27.
 */
class NateDetailApi(
    networkUseCase: NetworkUseCase
) : AbstractDetailApi(networkUseCase) {

    private val DETAIL_URL =
        "http://m.comics.nate.com/main2/webtoon/WebtoonView.php?btno=%s&bsno=%s"
    private lateinit var webToonId: String
    private lateinit var episodeId: String

    override fun parseDetail(episode: IEpisode): DetailResult {
        this.webToonId = episode.webToonId
        this.episodeId = episode.episodeId

        val ret = DetailResult.Detail(
            webtoonId = episode.webToonId,
            episodeId = episode.episodeId
        )

        val doc = try {
            Jsoup.parse(requestApi())
        } catch (e: Exception) {
            e.printStackTrace()
            return ret
        }
        ret.apply {
            title = doc.select(".tvi_header").text()

            doc.select(".btn_prev")?.apply {
                prevLink = if (isNotEmpty()) "/view/" + first().attr("href") else null
            }
            doc.select(".btn_next")?.apply {
                nextLink = if (isNotEmpty()) "/view/" + first().attr("href") else null
            }
            val list = mutableListOf<DetailView>()
            doc.select(".toonView img").mapTo(list) { DetailView(it.attr("src")) }
            this.list = list
        }
        return ret
    }

    override fun getDetailShare(episode: Episode, detail: DetailResult.Detail) = ShareItem(
        title = "${episode.title} / ${detail.title}",
        url = DETAIL_URL.format(episode.toonId, episode.episodeId)
    )

    override val method: REQUEST_METHOD = REQUEST_METHOD.GET

    override val url: String
        get() = DETAIL_URL.format(webToonId, episodeId)
}
